core/src/main/java/org/apache/james/core/User.java

```diff
@@ -0,0 +1,117 @@
+ *   http://www.apache.org/licenses/LICENSE-2.0                 *
+ *                                                              *
+ * Unless required by applicable law or agreed to in writing,   *
+ * software distributed under the License is distributed on an  *
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY       *
+ * KIND, either express or implied.  See the License for the    *
+ * specific language governing permissions and limitations      *
+ * under the License.                                           *
+ ****************************************************************/
+
+package org.apache.james.core;
+
+import java.util.Objects;
+import java.util.Optional;
+
+import com.google.common.base.MoreObjects;
+import com.google.common.base.Preconditions;
+
+public class User {
+    public static User fromUsername(String username) {
+        Preconditions.checkNotNull(username);
+
+        int domainDelimiterPosition = username.lastIndexOf('@');
+        if (domainDelimiterPosition == -1) {
+            return fromLocalPartWithoutDomain(username);
+        }
+        return fromLocalPartWithDomain(
+            username.substring(0, domainDelimiterPosition),
+            username.substring(domainDelimiterPosition + 1, username.length()));
+    }
+
+    public static User fromLocalPartWithDomain(String localPart, String domain) {
+        Preconditions.checkNotNull(domain);
+
+        return from(localPart,
+            Optional.of(domain));
+    }
+
+    public static User fromLocalPartWithoutDomain(String localPart) {
+        return from(localPart,
+            Optional.empty());
+    }
+
+    public static User from(String localPart, Optional<String> domain) {
+       return new User(localPart, domain);
+    }
+
+    private final String localPart;
+    private final Optional<String> domainPart;
+
+    private User(String localPart, Optional<String> domainPart) {
+        Preconditions.checkNotNull(localPart);
+        Preconditions.checkArgument(!localPart.isEmpty(), "username should not be empty");
+        Preconditions.checkArgument(!localPart.contains("@"), "username can not contain domain delimiter");
+
+        Preconditions.checkArgument(!isEmptyString(domainPart), "domain part can not be empty");
+        Preconditions.checkArgument(!containsDomainDelimiter(domainPart), "domain can not contain domain delimiter");
+
+        this.localPart = localPart;
+        this.domainPart = domainPart;
+    }
+
+    private Boolean containsDomainDelimiter(Optional<String> domainPart) {
+        return domainPart.map(s -> s.contains("@")).orElse(false);
+    }
+
+    private Boolean isEmptyString(Optional<String> domainPart) {
+        return domainPart.map(String::isEmpty)
+            .orElse(false);
+    }
+
+    public String getLocalPart() {
+        return localPart;
+    }
+
+    public Optional<String> getDomainPart() {
+        return domainPart;
+    }
+
+    public String getUsername() {
```

Comment:
it's not obvious that a username may contain the domain, there's probably a naming issue here

Reply:
IMO we should have a team rule saying **naming issue** should lead to **naming proposal**.

Reply:
looks a lot like a NaaS (naming as a service) : when I don't have a good name, I'll offload the choosing to team members.
As I can't find a good name, i propose to make domain field mandatory so that we can rename it to `asString()`. WDYT ?

Reply:
> Not OK for domain field mandatory as this object needs to be used in a virtual hosted, and non virtual hosted context.

How is these two subjects related ? In case you are in a virtual hosted context : domain is required. In the other case, we know the domain as there's a single one and then we can provide it when constructing the user.

So what's wrong ?